neuralpredictors/training/tracking.py

```diff
@@ -146,3 +153,360 @@ def asdict(self, time_key="time", make_copy=True):
+        (to be accumulated manually) or a callable (e.g. a function) that returns the objective value.
+
+        Args:
+            main_objective: key of the main objective that is used to e.g. decide lr reductions
+            **objectives: e.g. {"dataset": {"objective1": o_fct1, "objective2": 0, "normalization": 0},...}
+                           or {"dataset": {"task_key": {"objective1": o_fct1, "objective2": 0},...},...}
+        """
+        self.objectives = objectives
+        self.log = self._initialize_log(objectives)
+        self.time = []
+        self.main_objective = main_objective
+        self.epoch = -1
+        self.start_epoch()
+
+    def add_objectives(self, objectives: Dict, init_epoch: bool = False):
+        """
+        Add new objectives (with initial values) to the logs.
+        Args:
+            objectives: dictionary that needs to follow the same structure as self.log
+            init_epoch: flag that decides whether the a new epoch is initialized for this objective
+        """
+        deep_update(self.objectives, objectives)
+        new_log = self._initialize_log(objectives)
+        if init_epoch:
+            self._initialize_epoch(new_log, objectives)
+        deep_update(self.log, new_log)
+
+    def start_epoch(self):
+        """ Start a new epoch. Initialize each accumulation with its default value. """
+        t = time.time()
+        self.time.append(t)
+        self.epoch += 1
+        self._initialize_epoch(self.log, self.objectives)
+
+    def log_objective(self, value: float, key: Tuple[str, ...] = ()):
+        """
+        Add a new entry to the logs
+        Args:
+            value: objective score
+            key: hierarchical key to match in `self.log`
+        """
+        if key:
+            self._log_objective_value(value, self.log, key)
+        else:
+            self._log_objective_callables(self.log, self.objectives)
+
+    def display_log(self, key=(), tqdm_iterator=None):
+        """
+        Display the current objective value of everything under `key`
+        Args:
+            key: Tuple describing the objective to display.
+                 This could be something like `("Training","img_classification","loss")`
+                 to display the current classification loss or something like `("Training","img_classification")`
+                 to display everything we save for image classifcation training.
+            tqdm_iterator: A tqdm object that the log could be displayed on.
+        """
+        # normalize (if applicable) and turn into np.arrays:
+        n_log = self._normalize_log(self.log)
+        # flatten a subset of the dictionary:
+        current_log = self._gather_log(n_log, key, index=-1)
+        if tqdm_iterator:
+            tqdm_iterator.set_postfix(**current_log)
+        else:
+            logger.info(current_log)
+
+    def check_isfinite(self, log: dict = None) -> bool:
```

Comment:
I think the type annotation for `log` should be `Optional[Dict]`.

Reply:
Just realized that `log` can also be a Sequence. So the type annotation should be something like `Optional[Union[Sequence, Mapping]]`.